<a href="https://www.kaggle.com/code/youssef19/python-code-optimization-for-data-scientists?scriptVersionId=155858134" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# <center style="font-family: consolas; font-size: 32px; font-weight: bold;">  Write Efficient Python Code: Python Code Optimization for Data Scientists </center>
​
***



As a data scientist, your primary focus is extracting valuable insights from data, not idly waiting for code execution. The key to maximizing your impact lies in crafting efficient Python code that minimizes runtime and optimizes computational resources. In this comprehensive notebook, we embark on a journey to eliminate bottlenecks and rectify suboptimal design patterns, leveraging the prowess of Python's most revered libraries for data scientists: NumPy and pandas.

Discover how to transform your coding practices, streamline your workflow, and reclaim precious time for tasks that truly drive innovation. By delving into real-world examples and practical exercises, we'll navigate the intricacies of NumPy and pandas, empowering you to write code that not only runs faster but also unlocks the full potential of your data science endeavors.

Let's embark on this exploration together, unraveling the secrets of efficient Python coding and transforming the way you interact with data.

#### <a id="top"></a>
# <div style="box-shadow: rgb(60, 121, 245) 0px 0px 0px 3px inset, rgb(255, 255, 255) 10px -10px 0px -3px, rgb(31, 193, 27) 10px -10px, rgb(255, 255, 255) 20px -20px 0px -3px, rgb(255, 217, 19) 20px -20px, rgb(255, 255, 255) 30px -30px 0px -3px, rgb(255, 156, 85) 30px -30px, rgb(255, 255, 255) 40px -40px 0px -3px, rgb(255, 85, 85) 40px -40px; padding:20px; margin-right: 40px; font-size:30px; font-family: consolas; text-align:center; display:fill; border-radius:15px; color:rgb(60, 121, 245);"><b>Table of Contents</b></div>

**1. Making Your Code Efficient**

    1.1. Efficiently Combining, Counting, and Iterating

    1.2. Introduction to The Set Theory 

    1.3. Eliminating Loops

    1.4. Writing Better Loops
    

**2. Introduction to Pandas Optimization**

    2.1. Pandas DataFrame Iteration
    
    2.2. Pandas Alternative to Looping

    2.3. Optimal Pandas Iterating

# 1. Make Your Code Efficient

In this section, we will cover more efficiency tips and tricks. You will learn a few useful built-in modules for writing efficient code and practice using set theory. You'll then learn about looping patterns in Python and how to make them  more efficient.

## 1.1. Efficiently Combining, Counting, and Iterating

### Combining Objects

In this subsection, we'll cover combining, counting, and iterating over objects efficiently in python. Suppose we have two lists: one of the names and the other is the age for each of them. We want to combine these lists so that each name is stored next to its age. We can iterate over the names list using enumerate and grab each name's corresponding age using the index variable.

In [1]:
# combining objects 
names = ['Tauqeer', 'Javaid', 'Mohammed']
age = [25, 27, 40]
combined = []

for i,name in enumerate(names):
    combined.append((name, age[i]))
print(combined)

[('Ahmed', 25), ('Youssef', 27), ('Mohammed', 40)]


But Python's built-in function zip provides a more elegant solution. The name "zip" describes how this function combines objects like a zipper on a jacket (making two separate things become one). zip returns a zip object that must be unpacked into a list and printed to see the contents. Each item is a tuple of elements from the original lists.

In [2]:
# Combining objects with zip
combined_zip = zip(names, age)
print(type(combined_zip))

<class 'zip'>


In [3]:
combined_zip_list = [*combined_zip]
print(combined_zip_list)

[('Ahmed', 25), ('Youssef', 27), ('Mohammed', 40)]


### Counting Objects

Let's dig deeper into the Counter object. First, we will upload the pokemon dataset and print the first five rows of it then we will count the number of pokemon per each type using for loop and then using the Counter function.

Let's load the pokemon dataset and print the first five rows:

In [4]:
import pandas as pd
pokemon = pd.read_csv('../input/pokemon-dataset/pokemon.csv')
pokemon.head()


#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Now we will count the number of pokemon per each type using loops and calculate the execution time:

In [5]:
%%timeit -r1 -n1
# Counting with for loop
poke_types = pokemon['Type 1']
type_counts = {}
for poke_type in poke_types:
    if poke_type not in type_counts:
        type_counts[poke_type] = 1
    else:
        type_counts[poke_type] += 1
print(type_counts)

{'Grass': 70, 'Fire': 52, 'Water': 112, 'Bug': 69, 'Normal': 98, 'Poison': 28, 'Electric': 44, 'Ground': 32, 'Fairy': 17, 'Fighting': 27, 'Psychic': 57, 'Rock': 44, 'Ghost': 32, 'Ice': 24, 'Dragon': 32, 'Dark': 31, 'Steel': 27, 'Flying': 4}
4.06 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Finally, we will count the number of pokemon per each type using the Counter function and compare the time:

In [6]:
%%timeit -r1 -n1

# counting with collections.Counter()
from collections import Counter
poke_types = pokemon['Type 1']

type_counts = Counter(poke_types)
print(type_counts)

Counter({'Water': 112, 'Normal': 98, 'Grass': 70, 'Bug': 69, 'Psychic': 57, 'Fire': 52, 'Electric': 44, 'Rock': 44, 'Ground': 32, 'Ghost': 32, 'Dragon': 32, 'Dark': 31, 'Poison': 28, 'Fighting': 27, 'Steel': 27, 'Ice': 24, 'Fairy': 17, 'Flying': 4})
543 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


We can see that using the Counter function from the collections module is a more efficient approach. Just import Counter and provide the object to be counted. No need for a loop! Counter returns a Counter dictionary of key-value pairs. When printed, it's ordered by highest to lowest counts. If comparing runtime times, we'd see that using Counter takes less time compared to the standard dictionary approach!

## itertools

Another built-in module, itertools, contains functional tools for working with iterators. A subset of these tools is:
* Infinite iterators: count, cycle, repeat
* Finite iterators: accumulate, chain, zip_longest, etc.
* Combination generators: product, permutations, combinations

We will focus on combinatoric generators. These generators efficiently yield Cartesian products, permutations, and combinations of objects. Let's explore an example.

### Combinations 

Suppose we want to gather all combination pairs of Pokémon types possible. We can do this with a nested for loop that iterates over the poke_types list twice. Notice that a conditional statement is used to skip pairs having the same type twice. For example, if x is 'Bug' and y is 'Bug', we want to skip this pair. Since we're interested in combinations (where order doesn't matter), another statement is used to ensure either order of the pair doesn't already exist within the combos list before appending it. For example, the pair ('Bug', 'Fire') is the same as the pair ('Fire', 'Bug'). We want one of these pairs, not both.

In [7]:
#Combinations with loop

poke_types = ['Bug', 'Fire', 'Ghost', 'Grass', 'Water']
combos = []
for x in poke_types:
    for y in poke_types:
        if x == y:
            continue
        if ((x,y) not in combos) & ((y,x) not in combos):
            combos.append((x,y))
print(combos)

[('Bug', 'Fire'), ('Bug', 'Ghost'), ('Bug', 'Grass'), ('Bug', 'Water'), ('Fire', 'Ghost'), ('Fire', 'Grass'), ('Fire', 'Water'), ('Ghost', 'Grass'), ('Ghost', 'Water'), ('Grass', 'Water')]


The combinations generator from itertools provides a more efficient solution. First, we import combinations and then create a combinations object by providing the poke_types list and the length of combinations we desire. Combinations return a combinations object, which we unpack into a list and print to see the result. If comparing runtimes, we'd see using combinations is significantly faster than the nested loop.

In [8]:
# combinations with itertools
poke_types = ['Bug', 'Fire', 'Ghost', 'Grass', 'Water']
from itertools import combinations
combos_obj = combinations(poke_types, 2)
print(type(combos_obj))

combos = [*combos_obj]
print(combos)

<class 'itertools.combinations'>
[('Bug', 'Fire'), ('Bug', 'Ghost'), ('Bug', 'Grass'), ('Bug', 'Water'), ('Fire', 'Ghost'), ('Fire', 'Grass'), ('Fire', 'Water'), ('Ghost', 'Grass'), ('Ghost', 'Water'), ('Grass', 'Water')]


## 1.2. Introduction to The Set Theory

Often, we'd like to compare two objects to observe similarities and differences between their contents. When doing this type of comparison, it's best to leverage a branch of mathematics called the set theory. As you know, Python comes with a built-in set data type. Sets come with some handy methods we can use for comparing such as:
* intersection() : all elements that are in both sets
* difference(): all elements in one set but not the other
* symmetric_difference(): all elements in exactly one set
* union(): all elements that are in either set


When we'd like to compare objects multiple times and in different ways, we should consider storing our data in sets to use these efficient methods. Another nice feature of Python sets is their ability to quickly check if a value exists within its members. We call this membership testing using the in operator. We will go through how using  the in operator with a set is much faster than using it with a list or tuple.

Suppose we had two lists of Pokémon names: list_a and list_b and we would like to compare these lists to see which Pokémon appear in both lists. We could first use a nested for loop to compare each item in list_a to each item in list_b and collect only those items that appear in both lists. But, iterating over each item in both lists is extremely inefficient.

In [9]:
# Comparing objects with loops
list_a = ['Bulbasaur', 'Charmander', 'Squirtle']
list_b = ['Caterpie', 'Pidgey', 'Squirtle']

in_common = []
for pokemon_a in list_a:
    for pokemon_b in list_b:
        if pokemon_a == pokemon_b:
            in_common.append(pokemon_a)
print(in_common)

['Squirtle']


However, a better way is to use Python's set data type to compare these lists. By converting each list into a set, we can use the dot-intersection method to collect the Pokémon shared between the two sets. One simple line of code and no need for a loop!

In [10]:
# compaing objects using intersection()

list_a = ['Bulbasaur', 'Charmander', 'Squirtle']
list_b = ['Caterpie', 'Pidgey', 'Squirtle']
set_a = set(list_a)
set_b = set(list_b)
set_a.intersection(set_b)

{'Squirtle'}

We got the same answer with a much less number of code lines we can also compare the runtime to see how much using sets is much faster than using loops.

Compare the runtime between using sets and using for loops

In [11]:
%%timeit
in_common = []
for pokemon_a in list_a:
    for pokemon_b in list_b:
        if pokemon_a == pokemon_b:
            in_common.append(pokemon_a)

1.19 µs ± 8.63 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [12]:
%timeit in_common = set_a.intersection(set_b)

269 ns ± 6.56 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


We can see that using sets is much faster than using for loops. We can also use a set method to see Pokémon that exist in one set but not in another. To gather Pokémon that exist in set_a but not in set_b, use set_a.difference(set_b).

In [13]:
set_a = {'Bulbasaur', 'Charmander', 'Squirtle'}
set_b = {'Caterpie', 'Pidgey', 'Squirtle'}
set_a.difference(set_b)

{'Bulbasaur', 'Charmander'}

If we want the Pokémon in set_b, but not in set_a, we use set_b.difference(set_a).

In [14]:
set_a = {'Bulbasaur', 'Charmander', 'Squirtle'}
set_b = {'Caterpie', 'Pidgey', 'Squirtle'}
set_b.difference(set_a)

{'Caterpie', 'Pidgey'}

To collect Pokémon that exist in exactly one of the sets (but not both), we can use a method called the symmetric difference.

In [15]:
set_a = {'Bulbasaur', 'Charmander', 'Squirtle'}
set_b = {'Caterpie', 'Pidgey', 'Squirtle'}
set_a.symmetric_difference(set_b)

{'Bulbasaur', 'Caterpie', 'Charmander', 'Pidgey'}

Finally, we can combine these sets using the .union method. This collects all of the unique Pokémon that appear in either or both sets.

In [16]:
set_a = {'Bulbasaur', 'Charmander', 'Squirtle'}
set_b = {'Caterpie', 'Pidgey', 'Squirtle'}
set_a.union(set_b)

{'Bulbasaur', 'Caterpie', 'Charmander', 'Pidgey', 'Squirtle'}

Another nice efficiency gain when using sets is the ability to quickly check if a specific item is a member of a set's elements. Consider our collection of 720 Pokémon names stored as a list, tuple, and set.

In [17]:
names_list = list(pokemon['Name'])
names_set = set(pokemon['Name'])
names_tuple = tuple(pokemon['Name'])

We want to check whether or not the character, Zubat, is in each of these data structures and print the execution time for each data type:

In [18]:
%timeit 'Zubat' in names_list
%timeit 'Zubat' in names_tuple
%timeit 'Zubat' in names_set

864 ns ± 12.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
678 ns ± 4.83 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
60.7 ns ± 0.707 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


**When comparing runtimes, it's clear that membership testing with a set is significantly faster than a list or a tuple.**

One final efficiency gain when using sets comes from the definition of the set itself. A set is defined as a collection of distinct elements. Thus, we can use a set to collect unique items from an existing object. Let's define a primary_types list, which contains the primary types of each Pokémon. If we wanted to collect the unique Pokémon types within this list, we could write a for loop to iterate over the list, and only append the Pokémon types that haven't already been added to the unique_types list.

In [19]:
primary_types = list(pokemon['Type 1'])
unique_types = []
for prim_type in primary_types:
    if prim_type not in unique_types:
        unique_types.append(prim_type)
print(unique_types)

['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric', 'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice', 'Dragon', 'Dark', 'Steel', 'Flying']


Using a set makes this much easier. All we have to do is convert the primary_types list into a set, and we have our solution: a set of distinct Pokémon types.

In [20]:
unique_types_set = set(primary_types)
print(unique_types_set)

{'Ground', 'Normal', 'Fire', 'Electric', 'Ice', 'Grass', 'Bug', 'Steel', 'Poison', 'Rock', 'Dark', 'Ghost', 'Dragon', 'Fighting', 'Fairy', 'Psychic', 'Flying', 'Water'}


### 1.3. Eliminating Loops

Although using loops when writing Python code isn't necessarily a bad design pattern, using extraneous loops can be inefficient and costly. Let's explore some tools that can help us eliminate the need to use loops in our code. Python comes with a few looping patterns that can be used when we want to iterate over an object's contents:

* For loops iterate over elements of a sequence piece-by-piece. 
* While loops execute a loop repeatedly as long as some Boolean condition is met. 
* Nested loops use multiple loops inside one another.

Although all of these looping patterns are supported by Python, we should be careful when using them. Because most loops are evaluated in a piece-by-piece manner, they are often inefficient solutions.

We should try to avoid looping as much as possible when writing efficient code. Eliminating loops usually results in fewer lines of code that are easier to interpret. One of the idioms of pythonic code is that  "flat is better than nested." Striving to eliminate loops in our code will help us follow this idiom.

Suppose we have a list of lists, called poke_stats, that contains statistical values for each Pokémon. Each row corresponds to a Pokémon, and each column corresponds to a Pokémon's specific statistical value. Here, the columns represent a Pokémon's Health Points, Attack, Defense, and Speed. We want to do a simple sum of each of these rows in order to collect the total stats for each Pokémon. If we were to use a loop to calculate row sums, we would have to iterate over each row and append the row's sum to the totals list. We can accomplish the same task, in fewer lines of code, with a list comprehension. Or, we could use the built-in map function that we've discussed previously in the previous article.

In [21]:
# List of HP, Attack, Defense, Speed
poke_stats = [
[90, 92, 75, 60],
[25, 20, 15, 90],
[65, 130, 60, 75],
]

In [22]:
%%timeit
totals = []
for row in poke_stats:
    totals.append(sum(row))

953 ns ± 6.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [23]:
%timeit totals_comp = [sum(row) for row in poke_stats]

956 ns ± 9.66 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [24]:
%timeit totals_map = [*map(sum, poke_stats)]

790 ns ± 7.06 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


Each of these approaches will return the same list, but using a list comprehension or the map function takes one line of code, and has a faster runtime.

We've also covered a few built-in modules that can help us eliminate loops in the previous article. Instead of using the nested for loop, we can use combinations from the itertools module for a cleaner, more efficient solution.

In [25]:
poke_types = ['Bug', 'Fire', 'Ghost', 'Grass', 'Water']

# Nested for loop approach
combos = []
for x in poke_types:
    for y in poke_types:
        if x == y:
            continue
    if ((x,y) not in combos) & ((y,x) not in combos):
        combos.append((x,y))

In [26]:
# Built-in module approach
from itertools import combinations
combos2 = [*combinations(poke_types, 2)]

Another powerful technique for eliminating loops is to use the NumPy package. Suppose we had the same collection of statistics we used in a previous example but stored in a NumPy array instead of a list of lists.

We'd like to collect the average stat value for each Pokémon (or row) in our array. We could use a loop to iterate over the array and collect the row averages.

### Eliminate loops with NumPy

But, NumPy arrays allow us to perform calculations on entire arrays all at once. Here, we use the dot-mean method and specify an axis equal to 1 to calculate the mean for each row (meaning we calculate an average across the column values). This eliminates the need for a loop and is much more efficient.

In [27]:
import numpy as np

poke_stats = np.array([
[90, 92, 75, 60],
[25, 20, 15, 90],
[65, 130, 60, 75],])

In [28]:
avgs = []
for row in poke_stats:
    avg = np.mean(row)
    avgs.append(avg)
print(avgs)

[79.25, 37.5, 82.5]


In [29]:
avgs_np = poke_stats.mean(axis=1)
print(avgs_np)

[79.25 37.5  82.5 ]


In [30]:
%%timeit
avgs = []
for row in poke_stats:
    avg = np.mean(row)
    avgs.append(avg)

43.7 µs ± 355 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [31]:
%timeit avgs = poke_stats.mean(axis=1)

13.7 µs ± 219 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


When comparing runtimes, we see that using the dot-mean method on the entire array and specifying an axis is significantly faster than using a loop.

### 1.4. Writing Better Loops

We've discussed how loops can be costly and inefficient. But, sometimes you can't eliminate a loop. In this section, we'll explore how to make loops more efficient when looping is unavoidable. Before diving in, some of the loops we'll discuss can be eliminated using techniques covered in previous lessons. For demonstrative purposes, we'll assume the use cases shown here are instances where a loop is unavoidable.

The best way to make a loop more efficient is to analyze what's being done within the loop. We want to make sure that we aren't doing unnecessary work in each iteration. If a calculation is performed for each iteration of a loop, but its value doesn't change with each iteration, it's best to move this calculation outside (or above) the loop. If a loop is converting data types with each iteration, it's possible that this conversion can be done outside (or below) the loop using a map function. Anything that can be done once should be moved outside of a loop. Let's explore a few examples.

### Moving calculations above a loop

We have a list of Pokémon names and an array of each Pokémon's corresponding attack value. We'd like to print the names of each Pokémon with an attack value greater than the average of all attack values. To do this, we'll use a loop that iterates over each Pokémon and its attack value. For each iteration, the total attack average is calculated by finding the mean value of all attacks. Then, each Pokémon's attack value is evaluated to see if it exceeds the total average.

In [32]:
import numpy as np

names = ['Absol', 'Aron', 'Jynx', 'Natu', 'Onix']
attacks = np.array([130, 70, 50, 50, 45])

for pokemon,attack in zip(names, attacks):
    total_attack_avg = attacks.mean()
    if attack > total_attack_avg:
        print(
            "{}'s attack: {} > average: {}!"
            .format(pokemon, attack, total_attack_avg)
        )

Absol's attack: 130 > average: 69.0!
Aron's attack: 70 > average: 69.0!


The inefficiency in this loop is the total_attack_avg variable being created with each iteration of the loop. But, this calculation doesn't change between iterations since it is an overall average. We only need to calculate this value once. By moving this calculation outside (or above) the loop, we calculate the total attack average only once. We get the same output, but this is a more efficient approach.

In [33]:
import numpy as np

names = ['Absol', 'Aron', 'Jynx', 'Natu', 'Onix']
attacks = np.array([130, 70, 50, 50, 45])

# Calculate total average once (outside the loop)
total_attack_avg = attacks.mean()

for pokemon,attack in zip(names, attacks):
    if attack > total_attack_avg:
        print(
            "{}'s attack: {} > average: {}!"
            .format(pokemon, attack, total_attack_avg)
        )

Absol's attack: 130 > average: 69.0!
Aron's attack: 70 > average: 69.0!


Let's compare the runtimes of both methods:

In [34]:
import time

# get the start time
st = time.time()


names = ['Absol', 'Aron', 'Jynx', 'Natu', 'Onix']
attacks = np.array([130, 70, 50, 50, 45])

for pokemon,attack in zip(names, attacks):
    total_attack_avg = attacks.mean()
    if attack > total_attack_avg:
        print(
            "{}'s attack: {} > average: {}!"
            .format(pokemon, attack, total_attack_avg)
        )    
        
        
# get the end time
et = time.time()

# get the execution time
elapsed_time = et - st
print('Execution time:', elapsed_time, 'seconds')
        

Absol's attack: 130 > average: 69.0!
Aron's attack: 70 > average: 69.0!
Execution time: 0.0005266666412353516 seconds


In [35]:
import time

# get the start time
st = time.time()


# Calculate total average once (outside the loop)
total_attack_avg = attacks.mean()
for pokemon,attack in zip(names, attacks):
    if attack > total_attack_avg:
        print(
            "{}'s attack: {} > average: {}!"
            .format(pokemon, attack, total_attack_avg)
        )
        
# get the end time
et = time.time()

# get the execution time
elapsed_time = et - st
print('Execution time:', elapsed_time, 'seconds')
                

Absol's attack: 130 > average: 69.0!
Aron's attack: 70 > average: 69.0!
Execution time: 0.0003592967987060547 seconds


We see that keeping the total_attack_avg calculation within the loop takes more than 120 microseconds.

### Using holistic conversions

Another way to make loops more efficient is to use holistic conversions outside (or below) the loop. In the example below we have three lists from the 720 Pokémon dataset: a list of each Pokémon's name, a list corresponding to whether or not a Pokémon has a legendary status, and a list of each Pokémon's generation. We want to combine these objects so that each name, status, and generation is stored in an individual list. To do this, we'll use a loop that iterates over the output of the zip function. Remember, zip returns a collection of tuples, so we need to convert each tuple into a list since we want to create a list of lists as our output. Then, we append each individual poke_list to our poke_data output variable. By printing the result, we see our desired list of lists.

In [36]:
pokemon = pd.read_csv('../input/pokemon-dataset/pokemon.csv')

names_list = pokemon['Name']
legend_status_list = pokemon['Legendary']
generations_list = pokemon['Generation']

In [37]:
%%timeit -r1 -n1
poke_data = []
for poke_tuple in zip(names_list, legend_status_list, generations_list):
    poke_list = list(poke_tuple)
    poke_data.append(poke_list)

802 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


However, converting each tuple to a list within the loop is not very efficient. Instead, we should collect all of our poke_tuples together, and use the map function to convert each tuple to a list. The loop no longer converts tuples to lists with each iteration. Instead, we moved this tuple to list conversion outside (or below) the loop. That way, we convert data types all at once (or holistically) rather than converting in each iteration.

In [38]:
%%timeit -r1 -n1
poke_data_tuples = []
for poke_tuple in zip(names_list, legend_status_list, generations_list):
    poke_data_tuples.append(poke_tuple)
    poke_data = [*map(list, poke_data_tuples)]

60.8 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Runtimes show that converting each tuple to a list outside of the loop is more efficient.

# 2. Pandas Optimization

## 2.1. Pandas DataFrame Iteration

Pandas is a library used for data analysis. The main construct of pandas is the DataFrame, a tabular data structure with labeled rows and columns. We will focus on the best approaches for iterating over a DataFrame. Let's begin by analyzing a Major League Baseball dataset. It contains the team stats for each Major League Baseball team from the year 1962 to 2012, which are stored in a pandas DataFrame named baseball_df.

In [39]:
import pandas as pd
baseball_df = pd.read_csv('../input/baseball-stats/baseball_stats.csv')
baseball_df.head()

Team League  Year   RS   RA   W    OBP    SLG     BA  Playoffs  RankSeason  \
0  ARI     NL  2012  734  688  81  0.328  0.418  0.259         0         NaN   
1  ATL     NL  2012  700  600  94  0.320  0.389  0.247         1         4.0   
2  BAL     AL  2012  712  705  93  0.311  0.417  0.247         1         5.0   
3  BOS     AL  2012  734  806  69  0.315  0.415  0.260         0         NaN   
4  CHC     NL  2012  613  759  61  0.302  0.378  0.240         0         NaN   

   RankPlayoffs    G   OOBP   OSLG  
0           NaN  162  0.317  0.415  
1           5.0  162  0.306  0.378  
2           4.0  162  0.315  0.403  
3           NaN  162  0.331  0.428  
4           NaN  162  0.335  0.424

We will focus on the W column, which specifies the number of wins a team had in a season and the G column which contains the number of games a team played in a season.

### Calculating & Adding win percentage

One popular statistic used to evaluate a team's performance for a given season is the team's win percentage. This metric is calculated by dividing a team's total wins by the number of games played. Here is a simple function to perform this calculation.

In [40]:
def calc_win_perc(wins, games_played):
    win_perc = wins / games_played
    return np.round(win_perc,2)

Now we would like to create a new column in our baseball_df DataFrame that stores each team's win percentage for a season. To do this, we'll need to iterate over the DataFrame's rows and apply our calc_win_perc function. First, we create an empty win_perc_list to store all the win percentages we'll calculate. Then, we write a loop that will iterate over each row of the DataFrame. Notice that we are using an index variable (i) that ranges from zero to the number of rows that exist within the DataFrame. We then use the .iloc method to look up each individual row within the DataFrame using the index variable.

Then we grab each team's wins and games played by referencing the W and G columns. Next, we pass the team's wins and games played to calc_win_perc to calculate the win percentages. Finally, we append win_perc to win_perc_list and continue the loop. We create our desired column in the DataFrame, called WP, by setting the column value equal to the win_perc_list.

In [41]:
%%timeit
win_perc_list = []
for i in range(len(baseball_df)):
    row = baseball_df.iloc[i]
    wins = row['W']
    games_played = row['G']
    win_perc = calc_win_perc(wins, games_played)
    win_perc_list.append(win_perc)
baseball_df['WP'] = win_perc_list

baseball_df.head()

260 ms ± 2.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Looping over the DataFrame with .iloc gave us our desired output, but is it inefficient. When estimating the runtime, the .iloc approach took 660 milliseconds, which is pretty inefficient.

### Iterating with .iterrows()

pandas come with a few efficient methods for looping over a DataFrame. The first method we'll cover is the .iterrows method. This is similar to the .iloc method, but .iterrows returns each DataFrame row as a tuple of (index, pandas Series) pairs. This means each object returned from .iterrows contains the index of each row as the first element and the data in each row as a pandas Series as the second element.

Notice that we still create the empty win_perc_list, but now we don't have to create an index variable to look up each row within the DataFrame. .iterrows handles the indexing for us! The remainder of the for loop stays the same to create a new win percentage column within our baseball_df DataFrame.

In [42]:
%%timeit
win_perc_list = []
for i,row in baseball_df.iterrows():
    wins = row['W']
    games_played = row['G']

    win_perc = calc_win_perc(wins, games_played)
    win_perc_list.append(win_perc)

baseball_df['WP'] = win_perc_list

108 ms ± 2.46 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


Using .iterrows takes roughly half the time .iloc takes to iterate over our DataFrame.

### Iterating with .itertuples()

Pandas also come with another iteration method called .itertuples which is often more efficient than .iterrows. Let's continue using our baseball dataset to compare these two methods. Suppose we have a pandas DataFrame called team_wins_df that contains each team's total wins in a season.

If we use .iterrows to loop over our team_wins_df DataFrame and print each row's tuple, we see that each row's values are stored as a pandas Series. Remember, .iterrows returns each DataFrame row as a tuple of (index, pandas Series) pairs, so we have to access the row's values with square bracket indexing.

But, we could use .itertuples to loop over our DataFrame rows instead. The .itertuples method returns each DataFrame row as a special data type called a namedtuple. A namedtuple is one of the specialized data types that exist within the collections module we've discussed previously. These data types behave just like a Python tuple but have fields accessible using attribute lookup.

In [43]:
team_wins_df = baseball_df[['Team','Year','W']]
for row_namedtuple in team_wins_df.itertuples():
    print(row_namedtuple)

Pandas(Index=0, Team='ARI', Year=2012, W=81)
Pandas(Index=1, Team='ATL', Year=2012, W=94)
Pandas(Index=2, Team='BAL', Year=2012, W=93)
Pandas(Index=3, Team='BOS', Year=2012, W=69)
Pandas(Index=4, Team='CHC', Year=2012, W=61)
Pandas(Index=5, Team='CHW', Year=2012, W=85)
Pandas(Index=6, Team='CIN', Year=2012, W=97)
Pandas(Index=7, Team='CLE', Year=2012, W=68)
Pandas(Index=8, Team='COL', Year=2012, W=64)
Pandas(Index=9, Team='DET', Year=2012, W=88)
Pandas(Index=10, Team='HOU', Year=2012, W=55)
Pandas(Index=11, Team='KCR', Year=2012, W=72)
Pandas(Index=12, Team='LAA', Year=2012, W=89)
Pandas(Index=13, Team='LAD', Year=2012, W=86)
Pandas(Index=14, Team='MIA', Year=2012, W=69)
Pandas(Index=15, Team='MIL', Year=2012, W=83)
Pandas(Index=16, Team='MIN', Year=2012, W=66)
Pandas(Index=17, Team='NYM', Year=2012, W=74)
Pandas(Index=18, Team='NYY', Year=2012, W=95)
Pandas(Index=19, Team='OAK', Year=2012, W=94)
Pandas(Index=20, Team='PHI', Year=2012, W=81)
Pandas(Index=21, Team='PIT', Year=2012, W=79

Notice in the output that each printed row_namedtuple has an Index attribute and each column in our team_wins_df as an attribute. That means we can access each of these attributes with a lookup using a dot method. Here, we can print the last row_namedtuple's Index using **row_namedtuple.Index**. We can print this row_namedtuple's Team with **row_namedtuple.Team**, Year with **row_namedtuple.Year** and so on.

In [44]:
print(row_namedtuple.Index)

1231


In [45]:
print(row_namedtuple.Team)

WSA


When we compare **.iterrows** to **.itertuples**, we see that there is quite a bit of improvement!

In [46]:
%%timeit -r1 -n1
for row_tuple in team_wins_df.iterrows():
    print(row_tuple)

(0, Team     ARI
Year    2012
W         81
Name: 0, dtype: object)
(1, Team     ATL
Year    2012
W         94
Name: 1, dtype: object)
(2, Team     BAL
Year    2012
W         93
Name: 2, dtype: object)
(3, Team     BOS
Year    2012
W         69
Name: 3, dtype: object)
(4, Team     CHC
Year    2012
W         61
Name: 4, dtype: object)
(5, Team     CHW
Year    2012
W         85
Name: 5, dtype: object)
(6, Team     CIN
Year    2012
W         97
Name: 6, dtype: object)
(7, Team     CLE
Year    2012
W         68
Name: 7, dtype: object)
(8, Team     COL
Year    2012
W         64
Name: 8, dtype: object)
(9, Team     DET
Year    2012
W         88
Name: 9, dtype: object)
(10, Team     HOU
Year    2012
W         55
Name: 10, dtype: object)
(11, Team     KCR
Year    2012
W         72
Name: 11, dtype: object)
(12, Team     LAA
Year    2012
W         89
Name: 12, dtype: object)
(13, Team     LAD
Year    2012
W         86
Name: 13, dtype: object)
(14, Team     MIA
Year    2012
W         69
Name: 14, 

In [47]:
%%timeit -r1 -n1
for row_namedtuple in team_wins_df.itertuples():
    print(row_namedtuple)

Pandas(Index=0, Team='ARI', Year=2012, W=81)
Pandas(Index=1, Team='ATL', Year=2012, W=94)
Pandas(Index=2, Team='BAL', Year=2012, W=93)
Pandas(Index=3, Team='BOS', Year=2012, W=69)
Pandas(Index=4, Team='CHC', Year=2012, W=61)
Pandas(Index=5, Team='CHW', Year=2012, W=85)
Pandas(Index=6, Team='CIN', Year=2012, W=97)
Pandas(Index=7, Team='CLE', Year=2012, W=68)
Pandas(Index=8, Team='COL', Year=2012, W=64)
Pandas(Index=9, Team='DET', Year=2012, W=88)
Pandas(Index=10, Team='HOU', Year=2012, W=55)
Pandas(Index=11, Team='KCR', Year=2012, W=72)
Pandas(Index=12, Team='LAA', Year=2012, W=89)
Pandas(Index=13, Team='LAD', Year=2012, W=86)
Pandas(Index=14, Team='MIA', Year=2012, W=69)
Pandas(Index=15, Team='MIL', Year=2012, W=83)
Pandas(Index=16, Team='MIN', Year=2012, W=66)
Pandas(Index=17, Team='NYM', Year=2012, W=74)
Pandas(Index=18, Team='NYY', Year=2012, W=95)
Pandas(Index=19, Team='OAK', Year=2012, W=94)
Pandas(Index=20, Team='PHI', Year=2012, W=81)
Pandas(Index=21, Team='PIT', Year=2012, W=79

The reason **.itertuples** is more efficient than **.iterrows** is due to the way each method stores its output. Since .iterrows returns each row's values as a pandas Series, there is a bit more overhead.

## 2.2. Pandas Alternative to Looping

In order to write efficient code, we want to avoid looping as much as possible. Therefore now we will explore an alternative to using **.iterrows** and **.itertuples** to perform calculations on a DataFrame.

We'll continue using the  baseball dataset. We will first create the **calc_run_diff** function. This function calculates a team's run differential for a given year by subtracting the team's total number of runs allowed from their total number of runs scored in a season.

In [48]:
def calc_run_diff(runs_scored, runs_allowed):
    run_diff = runs_scored - runs_allowed
    return run_diff

We'd like to create a new column in our baseball_df DataFrame called RD that stores each team's run differentials over the years. Previously we did this with a for loop using either .iterrows or .itertuples. Here, we'll use .iterrows as an example. Notice that we are iterating over baseball_df with a for loop, passing each row's RS and RA columns into our calc_run_diff function, and then appending each row's result to our run_diffs_iterrows list. This gives us our desired output, but it is not our most efficient option.

In [49]:
%%timeit
run_diffs_iterrows = []

for i,row in baseball_df.iterrows():
    run_diff = calc_run_diff(row['RS'], row['RA'])
    run_diffs_iterrows.append(run_diff)

baseball_df['RD'] = run_diffs_iterrows

87.8 ms ± 1.31 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


### Pandas.apply() Method

One alternative to using a loop to iterate over a DataFrame is to use pandas.apply method. This function acts like the map function. It takes a function as an input and applies this function to an entire DataFrame. Since we are working with tabular data, we must specify an axis that we would like our function to act on. Using a zero for the axis argument will apply our function on columns while using a one for the axis will apply our function on all rows. Just like the map function, pandas.apply method can be used with anonymous functions or lambdas. 

Let's walk through how'd we'd use the .apply method to calculate our run differentials. First, we call .apply on the baseball_df DataFrame. Then, we use a lambda function to iterate over the rows of the DataFrame. Notice that our argument for lambda is row (since we are applying to each row of the DataFrame). For every row, we grab the RS and RA columns and pass them to our calc_run_diff function. Lastly, we specify our axis to tell dot-apply that we want to iterate over rows instead of columns.

In [50]:
%%timeit
run_diffs_apply = baseball_df.apply(
    lambda row: calc_run_diff(row['RS'], row['RA']),
    axis=1)

baseball_df['RD'] = run_diffs_apply

30.7 ms ± 361 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


When we use the .apply method to calculate our run differentials, we don't need to use a for a loop. We can collect our run differentials directly into an object called run_diffs_apply. 

We can see that the .iterrows approach took about 321 milliseconds to complete. But, using the dot-apply method took only 45.2 milliseconds. A definite improvement!

### 2.3. Optimal Pandas Iterating

We've come a long way from our first .iloc approach for iterating over a DataFrame. Each approach we've discussed has really improved the performance. But, these approaches focus on performing calculations for each row of our DataFrame individually. Now we will explore some panda's internals that allows us to perform calculations more efficiently.

Since pandas is a library that is built on NumPy. This means that each pandas DataFrame we use can take advantage of the efficient characteristics of NumPy arrays discussed in the previous article. We discussed broadcasting and how it allows the NumPy arrays to vectorize operations, so they are performed on all elements of an object at once. This allows us to efficiently perform calculations over entire arrays. Just like NumPy, pandas are designed to vectorize calculations so that they operate on entire datasets at once (not just on a row-by-row basis). Let's explore this concept with some examples.

We'll continue to use the baseball_df DataFrame we have been using throughout the chapter. Since pandas are built on top of NumPy, we can grab any of these DataFrame column's values as a NumPy array using the dot-values method. Here, we are collecting the W column's values into a NumPy array called wins_np. When we print the type of wins_np, we see that it is, in fact, a NumPy array.

In [51]:
wins_np = baseball_df['W'].values
print(type(wins_np))

<class 'numpy.ndarray'>


We can see the contents of the array by printing it and verifying that it is the same as the W column from our DataFrame.

In [52]:
print(wins_np)

[ 81  94  93 ... 103  84  60]


The beauty of knowing that pandas are built on NumPy can be seen when taking advantage of a NumPy array's broadcasting abilities. Remember, this means we can vectorize our calculations, and perform them on entire arrays all at once! Instead of looping over a DataFrame, and treating each row independently, like we've done with .iterrows, .itertuples, and .apply, we can perform calculations on the underlying NumPy arrays of our baseball_df DataFrame. Here, we gather the RS and RA columns in our DataFrame as NumPy arrays, and use broadcasting to calculate run differentials all at once!

In [53]:
%%timeit

run_diffs_np = baseball_df['RS'].values - baseball_df['RA'].values
baseball_df['RD'] = run_diffs_np

162 µs ± 2.74 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


When we use NumPy arrays to perform our run differential calculations, we can see that our code becomes much more readable. Here, we can explicitly see how our run differentials are being calculated. Also when we time our NumPy arrays approach, we see that our run differential calculations take microseconds! All other approaches were reported in milliseconds. Our array approach is orders of magnitude faster than all previous approaches! It's clear that using a DataFrame's underlying NumPy arrays to perform calculations can help us gain some massive efficiencies.